src/fennel/view.fnl

```diff
@@ -255,6 +255,7 @@ as numeric escapes rather than letter-based escapes, which is ugly."
           ;; main serialization loop, entry point is defined below
           (let [indent (or indent 0)
                 options (or options (make-options x))
+                x (if options.preprocess (options.preprocess x options) x)
```

Comment:
Have you considered for `prerocess` not to be a function, but a `preprocessors` table?

``` clojure
{:key (fn [ud] ...)
 :other-type (fn [ud] ...)}
```

This way way, we will be able to specify a conversion function tied to a particular type. Then, `pp` function instead of converting everything using the same function would instead call `type` on the `x` and search if the resulting type exists in `options.preprocessors`:

``` clojure
(let [indent (or indent 0)
      options (or options (make-options x))
      x (match (. options.preprocessors (type x))
          f (f x options)
          _ x)
      tv (type x)]
  (if (or (= tv :table)
...
```

I think this will allow a more fine control over userdata printing, and will help avoid confusion and errors when all data is preprocessed for no real reason.

Reply:
> I think this will allow a more fine control over userdata printing

How is it better than a single `if` within the function itself?

What if I want to trim `:password` field from _all_ tables, regardless how nested they are?

> will help avoid confusion and errors when all data is preprocessed for no real reason

I think a better way to do that would be to give an example of the `preprocess` function with a clear default return value with a comment describing why it's done that way.

Reply:
>
> How is it better than a single if within the function itself?
>
This provides an interface. And it makes function far less complected - the
type matching is not function's job, the data transformation is.

> What if I want to trim :password field from *all* tables, regardless how
> nested they are?

``` clojure
{:key (fn [ud] ...)
 :table (fn [t] (hide-password ...)}
```

This will transform any tables regardless of how nested they are, and as a
bonus you will not need to modify an existing function for awesome wm key
userdata transformer.

> will help avoid confusion and errors when all data is preprocessed for no
> real reason
>
> I think a better way to do that would be to give an example of the
> preprocess function with a clear default return value with a comment
> describing why it's done that way.
>
The clear default for preprocess function is the value it was given. This
is the same as not applying any function at all.

5 simple functions working with their 5 respecting data types is less
complect than 1 function working with 5 data types.

>

Reply:
I think adding this option is a good idea as an alternative when `__fennelview` either can't be used or would be unwieldy. @p-himik do you mind adding a couple test cases? It's just a one-line change, but we'll want some coverage for it just to prevent any future regressions and the like.

### preprocess fn vs preprocessors table
I would prefer the the `options.preprocess` as written over cooking in the type dispatching:

- It keeps the preprocessors composable - if you want to combine preprocessing logic you can just compose functions together instead of composing multiple functions in multiple tables
- The type dispatching in the function doesn't really add that much complexity if you need it (it's a very common, and clean, usecase of `match`), while not locking you into any specific patterns (you could also just do duck-typing).
- A preprocessor that appllies to multiple types and/or needs to do some sort of duck-typing or other conditional logic would take at least as much extra boilerplate with a type-dispatching preprocessors table as keeping the option simple and leaving type dispatch to the function itself.